### ISSUE TITLE
Add refactoring action to annotate a variable with its inferred type

### ISSUE THREAD

**Issue by rotemdan** [Microsoft/TypeScript#13199]
Hi, I'm using latest stable VSCode and I'm not sure if codefixes and refactorings have been implemented there yet, in any case, having a refactoring option to take any variable which receives an implicit inferred type, say, through a function call:

```ts
function test(): number[] {
  // ...
}

let x = test(); 
```
And being able to right-click `x` and invoke an action like `Annotate with inferred type` (or `Extract inferred type`), such that the resulting statement would look like:

```ts
let x: number[] = test();
```

could be very useful!

**Comment by aluanhaddad** [comment]
I would certainly appreciate having the opposite capability.

**Comment by mhegazy** [comment]
Marking as "Awaiting More Feedback". Would love to hear from others what are the expected experience here.

**Comment by rotemdan** [comment]
@aluanhaddad and others who showed support.

I taking it you're serious about this, otherwise I have don't really know what you mean here, and I suggest adding more details. 

I agree it would be useful to have the opposite capability as well, where annotated types that match their inferred types can be safely removed, but if that action is available, it seems natural to have the one I suggested for completion anyway.

@mhegazy 

This is just a tool suggestion, not a language feature, so it doesn't really have any impact on any particular "experience" (I guess either it's there or not). The experience is that the user can right click an individual identifier within a `let`, `var` or `const` statement, choose `Annotate with inferred type` and "document" the type into the code. There are cases where I find this useful, especially when the type is complex, or I want to "freeze" the type of a variable such that any change in the origin of the inferred type (function return type, secondary variable, literal) would mark the declared type as mismatched, rather than having various random subtle errors scattered all over the code, that are harder to interpret.

There are cases where I would use the opposite functionality as well, I guess, perhaps more like a global action, per file or per selection. I think this is a personal style choice, and depends on the particular scenario.

**Comment by rotemdan** [comment]
Also, being able to annotate functions or methods with their inferred return types, e.g.

```ts
function test() {
  return [1,2,3,4]
}
```
Right click `test`, select `Annotate with inferred return type, leading to:

```ts
function test(): number[] {
  return [1,2,3,4]
}
```

Would also be very useful. I do that a lot, sometimes to make sure that the inferred return type is stabilized such that further changes to the body of the function that break it would be detected.

There are also cases where callback's parameter types are inferred like

```ts
function example(f: (a: number, b: string) => boolean) {
}

example((x, y) => {
})
```
Where right clicking `x` or `y` and choosing `Annotate with inferred type` could also be useful.

**Comment by gcnew** [comment]
It would be awesome if there are both the array and tuple inference to choose from in the (small heterogenous) array case. Currently it's very irritating to manually annotate them..

**Comment by aluanhaddad** [comment]
@rotemdan 
> I agree it would be useful to have the opposite capability as well, where annotated types that match their inferred types can be safely removed, but if that action is available, it seems natural to have the one I suggested for completion anyway.

I agree, I just worry that there will be some cultures where it becomes the established practice to run this refactoring as matter of course.
A big negative result of this is that type inference is short circuited and improvements and refinements to types no longer propagate to their use sites.
Another thing that comes to mind is that it may mean a very different thing to freeze the type of a value typed by an explicitly declared type as compared to freezing the type of a value typed by a "synthesized" union or intersection that has resulted from some confluence of inferences.
I am not saying this is a bad idea, but it would be important to use it prudently.

**Comment by rotemdan** [comment]
@aluanhaddad 

This is just a utility, not a language feature, as I've already said, it cannot "degrade" the experience of the language or even the editor. It is not the role of the authors of refactoring tools to try to manipulate what users should or should do with the utilities they provide. I also suggested the annotation as an individual action, not a global one so that part of the process is that the inferred type is visually validated by the developer.

The funny thing is that the "remove unused variables" action is potentially more dangerous, as it can accidentally cause the loss of intermediate work that's in progress (though I personally don't see any problem with having it, again, it's just a utility, no one is inclined to use it). For some reason though people decided to antagonize this one.

**Comment by rotemdan** [comment]
Here's an enhancement idea: maybe instead of using the text like `Annotate with inferred type`, the tool will immediately show the inferred type in the menu option:

Something like:

```ts
function func(): number[];

let x = func()
```
Right clicking `x` the menu option would look like "Annotate with `number[]`", so now the programmer can inspect and validate the type even before it is added to the code.

Similarly:
```ts
function func() {
  return ["a", "b", "c"]
}
```
Right clicking `func` the menu option would look like `Annotate with () => string[]`.
Possibly also to address @gcnew's request, it would also include the option `Annotate with () => [string, string, string]`.

**Comment by rozzzly** [comment]
What about `['a', 'b', 'c']`?

**Comment by jcalz** [comment]
This would be helpful to me, also.  One use case I have is being able to extract an interface from a private or anonymous class so it can be referenced in other scopes.  

For example, inner classes (that is, classes as non-static properties of an _instance_ of an outer class) can't be easily statically typed:
```ts

class OuterClass {
  InnerClass: { new(...args: any[]): any }; // would like something more specific here
  outerProp: string;
  outerMethod() {
    return this.outerProp;
  }
  constructor(outerProp: string) {
    this.outerProp = outerProp;
    var outerInstance = this;
    this.InnerClass = class InnerClass {
      outerInstance = outerInstance;
      innerProp: string;
      constructor(innerProp: string) {
        this.innerProp = innerProp;
      }
      innerMethod() {
        return outerInstance.outerProp + ': ' + this.innerProp;
      }
    };
  }
}

var outerInstance = new OuterClass('howdy');
console.log(outerInstance.outerMethod());
var innerInstance = new outerInstance.InnerClass(0xd00d1e); // oops, bad argument
console.log(innerInstance.innerMethud()); // oops, bad method name
```

It would be nice if I go to the `class InnerClass` line and extract interfaces for the instance and static sides, such as:
```ts
interface InnerClass {
  outerInstance: OuterClass;
  innerProp: string;
  innerMethod: () => string;
}
interface InnerClassConstructor {
  new(innerProp: string): InnerClass;
}
class OuterClass {
  InnerClass: InnerClassConstructor;
⋮
```
which would catch the errors below.  Of course, one can manually examine the class definition and pull an interface out of it (as I did), but it can be tedious and error-prone.   It would be amazing if the tooling could tell us what structural type the type checker thinks an expression is, and either annotate the expression with that type or let us copy the type somewhere else.

**Comment by RyanCavanaugh** [comment]
@jcalz I believe editing the comment triggered GitHub to "fix" an invalid assignment (zhengbli is no longer a contributor). So no worries.

Broader refactoring support is on the way and things like this will be easy for us to add, so I expect we might do this in the future.

**Comment by safareli** [comment]
This feature is gonna be great improvement of developer experience. In PureScript for example if a top level declaration has no type annotation compiler gives a warning and using [purescript-ide-vscode](https://github.com/nwolverson/vscode-ide-purescript) extension user can perform quick fix on such warning and type annotation will be added to to the declaration instantly.

**Comment by safareli** [comment]
If anyone knows an extension which provides such functionality please comment here

**Comment by nick-lvov-dev** [comment]
Hey guys, just wanted to mention, I had the same issue and made an extension for this.
Check it out and send me your feedback!
https://marketplace.visualstudio.com/items?itemName=nick-lvov-dev.typescript-explicit-types

**Comment by Richiban** [comment]
This is a refactoring I would love to have, since I most of the time write functions and methods without explicit return types but later, when the design of the code has settled, I like to "pin" the types of exported functions so that they don't change without my explicit say-so.

For example:

```ts
function utilFunction() { ... }

export function doImportantThing() {
    return utilFunction();
}
```

In the given example, changing the body of `utilFunction` may well change its return type, in which case the return type of `doImportantThing` will change as well. During rapid development this is a desirable feature but, as the codebase gets big, I like to explicitly type the returns of important functions. My process for doing this at the moment is to hover over the function to see the referred return type and then I essentially just copy it right out of the tooltip there. It would be great if there was a way I could trigger this automatically.
